# Week 4 Assignment 1 - 'To-do 리스트 Redux를 사용하여 리팩터링 하기'

SangYoonLee1231 opened the pull request:
###  🙇‍♀️ 인사
- 안녕하세요! 저번 주 학습이 부진했던 것이 마음에 걸려 오늘은 첫 날부터 PR을 보냅니다.
- 그렇지만 아직 기본 세팅 작업밖에 하지 않았고, 몸도 다 나은 것이 아니라
  과제를 수행하려면 시간이 좀 더 필요할 것 같습니다.
- 이번에도 잘 부탁드립니다! 👊
 
<br/>

### ☑ 과제 구현 목표
- To-do 앱의 상태관리를 Redux로 변경하여 리펙토링을 진행한다.
- TDD로 모든 기능을 구현해야 한다.
- 모든 테스트가 통과해야 한다. (커버리지 100% 달성)

<br/>

### 📄 과제 수행 계획
- [x] <strong>1차 목표</strong>. ToDo 앱에서 State를 분리하여 Redux로 관리하기
- [x] <strong>2차 목표</strong>. <code>reducer</code> 테스트 코드 작성
- [ ] <strong>3차 목표</strong>. 컴포넌트 구조 변경하기
- [ ] <strong>4차 목표</strong>. 리펙토링을 통한 코드 개선

<br/>

### ✏ 과제 진행
#### 1차 제출
- 기본 세팅 작업

#### 2차 제출
- Redux 개인 학습을 진행한 후, 코드숨 온라인 강의를 보며  
  우선 ToDo 앱에서 state를 분리하여 이를 Redux로 관리하도록 구현했습니다. (8/24)
- 강의를 보며 reducer 테스트 코드 작성까지 완료했습니다. (8/25)

<br/>

SangYoonLee1231 committed:
First Submit

- 기본 세팅 완료

daadaadaah commented:
선PR 너무 좋습니당!👍 몸이 아직 다 안나으셨지만, 이번주 같이 힘내서 화이팅해봐요👊

SangYoonLee1231 committed:
Second Submit

- Redux 개인 학습 후, 코드숨 온라인 강의를 보며 ToDo 앱에서 state를 분리한 후 이를 Redux로 관리하도록 구현하였습니다.

SangYoonLee1231 committed:
Separate state management by redux and Add reduer Test

- 강의를 보며 reducer 테스트 코드 작성까지 완료했습니다.

SangYoonLee1231 commented:
- 저번 주 Jest 까지는 어떻게든 따라갈 수 있었는데, 이번 주 Redux는 특히 강의 코드를 이해하는 것이 너무 어렵네요..
- Redux의 개념은 생활 코딩 강의 + velopart님 강의 문서를 통해 어느 정도 이해를 했습니다만..

daadaadaah commented on src/index.jsx:
```diff
@@ -1,8 +1,15 @@
+import React from 'react';
 import ReactDOM from 'react-dom';
-
+import { Provider } from 'react-redux';
 import App from './App';
```
맥락 구분을 위해 한줄 띄어줍시당!

daadaadaah commented on src/store.js:
```diff
@@ -0,0 +1,6 @@
+import { createStore } from 'redux';
+import reducer from './reducer';
```
맥락 구분을 위해 한줄 띄어줍시당!

daadaadaah commented on src/reducer.js:
```diff
@@ -0,0 +1,38 @@
+};
+
+export default function reducer(state = initialState, action = '') {
+  if (action.type === 'updateTask') {
+    return {
+      ...state,
+      newTask: action.payload.newTask,
+    };
+  }
+
+  if (action.type === 'addTask') {
+    const { newId, newTask, tasks } = state;
+
+    if (!newTask) {
+      return state;
+    }
+
+    return {
+      ...state,
+      newId: newId + 1,
+      tasks: [...tasks, { id: newId, title: newTask }],
+    };
+  }
+
+  if (action.type === 'deleteTask') {
+    const { tasks } = state;
+    return {
+      ...state,
+      tasks: tasks.filter((task) => task.id !== action.payload.id),
+    };
+  }
```
if 문이 반복되고 있어요! 어떻게 하면 if 문을 사용하지 않고 reducer를 구현할 수 있을까요? 🤔

SangYoonLee1231 replied:
음.. 맞는 방법인진 모르겠지만 switch문을 이용할 수 있을 것 같습니다..!
그렇게 리펙토링 해보겠습니다!

daadaadaah replied:
과제 제한 조건에 switch를 사용하지 말라고 있습니다!ㅎㅎ1주차 때, if-else 문이나 switch 대신 아샬님은 어떻게 문제를 해결하셨는지 잘 되새겨봅시당!ㅎㅎ

SangYoonLee1231 replied:
앗 깜박했습니다..😅
음.. 그렇다면 매개변수나 객체를 이용해서 해결해야 할 것 같습니다!

그렇게 고민을 좀 해보았는데.. 객체를 활용해야 할 것 같은 감은 오지만, 코드를 어떻게 리펙토링을 해야할 진 잘 모르겠습니다.. 😥

daadaadaah replied:
오! 객체를 활용해야 할 것 같다는 생각을 하신거 잘하셨어요!👏
코드 일부만 보여드리면, 이런 형태가 나오기 위해서 reducers 객체를 어떻게 만드는 것이 좋을까?를 한번 생각해보세요!
```js
const reducer = (state = initialState, action) => {
  return reducers[action.type](state, action);
};
```

daadaadaah reviewed (changes requested):
1. 다른 자료들 보면서까지 이해하려고 하는 모습 너무 좋습니다!👍 특히 어떤 부분이 어렵다고 느끼셨나요? 저도 아샬님 강의 들었을 때를 생각해보며 도움을 드릴 수 있을까 해서요!ㅎㅎ
2. TDD로 작업하시면서 작업 내역 쪼개가며 commit 메시지 남겨주시는거 잊지마시구~ 다음 PR 기다리고 있을게요! 👊

daadaadaah commented on src/store.js:
```diff
@@ -0,0 +1,6 @@
+import { createStore } from 'redux';
```
혹시, createStore()가 deprecated 되었다고 안뜨나용?ㅎㅎ

SangYoonLee1231 replied:
앗 뜹니다! 다른 분 PR을 참고하니 configureStore로 진행해달라고 말씀하셨군요..!
그렇게 해보겠습니다!

SangYoonLee1231 commented:
> 1. 다른 자료들 보면서까지 이해하려고 하는 모습 너무 좋습니다!👍 특히 어떤 부분이 어렵다고 느끼셨나요? 저도 아샬님 강의 들었을 때를 생각해보며 도움을 드릴 수 있을까 해서요!ㅎㅎ

사실 야살님의 첫 번째 강의 중 App 테스트 코드에서 가짜로 useSelector을 써서 쓰는 부분이 (27:40부터 입니다) 잘 이해가 되지 않았었는데, 오늘 다시 보고 이해했습니다. ㅎㅎ

아쉽게도 오늘은 개인 사정으로 인해 PR을 남기지 못했네요ㅠ
내일 해설강의 보면서 마저 과제 수행하고 PR 드리겠습니다! 🙂

SangYoonLee1231 committed:
Fix : 리뷰 반영하여 수정

- 맥락 구분을 위한 띄어쓰기 반영
- createStore을 configureStore로 수정

- if문을 리펙토링하는 건 잘 모르겠어서 아직 수정하지 못하였습니다..
